pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -3647,7 +3647,14 @@ public CompletableFuture<Position> getLastDispatchablePosition() {
             return ManagedLedgerImplUtils.asyncGetLastValidPosition((ManagedLedgerImpl) ledger, entry -> {
                 MessageMetadata md = Commands.parseMessageMetadata(entry.getDataBuffer());
                 // If a messages has marker will filter by AbstractBaseDispatcher.filterEntriesForConsumer
-                return !Markers.isServerOnlyMarker(md);
+                if (Markers.isServerOnlyMarker(md)) {
+                    return false;
+                } else if (md.hasTxnidMostBits() && md.hasTxnidLeastBits()) {
+                    // Filter-out transaction aborted messages.
+                    TxnID txnID = new TxnID(md.getTxnidMostBits(), md.getTxnidLeastBits());
+                    return !isTxnAborted(txnID, (PositionImpl) entry.getPosition());
```

Comment:
This will search forward one by one until a valid position is found, this performance very low if the transaction has a lot of abort messages.

Reply:
@coderzc Yes, But we are unable to do more optimizations, TransactionBuffer didn't record aborted messages' position.

Reply:
Same in https://github.com/apache/pulsar/pull/21951.

Reply:
@thetumbled @liangyepianzhou It's a further improvement, I'm going to read more entries from ledger(not one by one) to optimize it.